front/src/modules/timesStops/helpers/utils.ts

```diff
@@ -125,3 +139,145 @@ export function disabledTextColumn(
     disabled: true,
   };
 }
+
+/**
+ * Synchronizes arrival, departure and stop times.
+ * updates onStopSignal
+ * updates isMarginValid and theoreticalMargin
+ */
+export function updateRowTimesAndMargin(
+  rowData: PathWaypointRow,
+  previousRowData: PathWaypointRow,
+  op: { fromRowIndex: number },
+  allWaypointsLength: number
+): PathWaypointRow {
+  const newRowData = { ...rowData };
+  if (
+    newRowData.departure &&
+    newRowData.arrival &&
+    (!isEqual(newRowData.arrival, previousRowData.arrival) ||
+      !isEqual(newRowData.departure, previousRowData.departure))
+  ) {
+    newRowData.stopFor = String(
+      durationInSeconds(
+        time2sec(newRowData.arrival.time || ''),
+        time2sec(newRowData.departure.time || '')
+      )
+    );
+  }
+  if (!newRowData.stopFor && op.fromRowIndex !== allWaypointsLength - 1) {
+    newRowData.onStopSignal = false;
+  }
+  newRowData.isMarginValid = !(
+    newRowData.theoreticalMargin && !marginRegExValidation.test(newRowData.theoreticalMargin)
+  );
+  if (newRowData.isMarginValid && op.fromRowIndex === 0) {
+    newRowData.arrival = undefined;
+    // As we put 0% by default for origin's margin, if the user removes a margin without
+    // replacing it to 0% (undefined), we change it to 0%
+    if (!newRowData.theoreticalMargin) {
+      newRowData.theoreticalMargin = '0%';
+    }
+  }
+  return newRowData;
+}
+
+/**
+ * This function goes through the whole array of path waypoints
+ * and updates the number of days since departure.
+ */
+export function updateDaySinceDeparture<T extends Pick<PathWaypointRow, 'arrival' | 'stopFor'>>(
+  pathWaypointRows: T[],
+  startTime?: IsoDateTimeString,
+  keepFirstIndexArrival?: boolean
+): T[] {
```

Comment:
Why can't we just do it like that since everywhere we use this function we pass it a `PathWaypointRow` array ?

```suggestion
export function updateDaySinceDeparture(
  pathWaypointRows: PathWaypointRow[],
  startTime?: IsoDateTimeString,
  keepFirstIndexArrival?: boolean
): PathWaypointRow[] {
```

Reply:
done; the function received two kind of object with arrival and departure, but it’s changed since the store changed.